segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/GarbageCollector.java

```diff
@@ -132,240 +160,339 @@ public GarbageCollector(int containerId, ChunkStorage chunkStorage,
+     * Adds given chunks to list of garbage chunks.
      *
-     * @return The Duration.
+     * @param chunksToDelete List of chunks to delete.
      */
-    @Override
-    protected Duration getShutdownTimeout() {
-        return SHUTDOWN_TIMEOUT;
+    CompletableFuture<Void> addChunksToGarbage(long transactionId, Collection<String> chunksToDelete) {
+        if (null != taskQueue) {
+            val currentTime = currentTimeSupplier.get();
+            val futures = new ArrayList<CompletableFuture<Void>>();
+            chunksToDelete.forEach(chunkToDelete -> futures.add(addChunkToGarbage(transactionId, chunkToDelete, currentTime + config.getGarbageCollectionDelay().toMillis(), 0)));
+            return Futures.allOf(futures);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
     /**
-     * Main execution of the Service. When this Future completes, the service auto-shuts down.
+     * Adds given chunk to list of garbage chunks.
      *
-     * @return A CompletableFuture that, when completed, indicates the service is terminated. If the Future completed
-     * exceptionally, the Service will shut down with failure, otherwise it will terminate normally.
+     * @param chunkToDelete Name of the chunk to delete.
+     * @param startTime Start time.
+     * @param attempts Number of attempts to delete this chunk so far.
      */
-    @Override
-    protected CompletableFuture<Void> doRun() {
-        loopFuture = Futures.loop(
-                this::canRun,
-                () -> delaySupplier.get()
-                        .thenComposeAsync(v -> deleteGarbage(true, config.getGarbageCollectionMaxConcurrency()), executor)
-                        .handleAsync((v, ex) -> {
-                            if (null != ex) {
-                                log.error("{}: Error during doRun.", traceObjectId, ex);
-                            }
-                            return null;
-                        }, executor),
-                executor);
-        return loopFuture;
+    CompletableFuture<Void>  addChunkToGarbage(long transactionId, String chunkToDelete, long startTime, int attempts) {
+        if (null != taskQueue) {
+            return taskQueue.addTask(taskQueueName, new TaskInfo(chunkToDelete, startTime, attempts, TaskInfo.DELETE_CHUNK, transactionId))
+                    .thenRunAsync(() -> {
+                        queueSize.incrementAndGet();
+                        SLTS_GC_CHUNK_QUEUED.inc();
+                    }, this.storageExecutor);
+        }
+        return CompletableFuture.completedFuture(null);
     }
 
-    private boolean canRun() {
-        return isRunning() && getStopException() == null && !closed.get();
+    CompletableFuture<Void>  addSegmentToGarbage(long transactionId, String segmentToDelete) {
+        if (null != taskQueue) {
+            val startTime = currentTimeSupplier.get() + config.getGarbageCollectionDelay().toMillis();
```

Comment:
I think I have seen the behavior of this reflected in the metrics of our experiments. This value, which is 60 seconds if I'm not mistaken, is just a fixed delay for work that needs to be done anyway, right? My question is: why is this the right behavior? Why do we need to buffer in the queue 60 seconds worth of tasks before start processing them? Can this lead to overloading the queue under a spike of delete tasks?

Reply:
We want to delay process deletion logic (currently after 1 minute) that is the whole point of having GC and not do this work inline. Each SLTS operation must be atomic - it can not have partial results (more specifically effect that are non idempotent)

This particular behavior is desired new behavior - to break down segment deletion into multiple steps.

First step will just mark segment inactive and enqueue GC task - this way delete is fast for storage writer and doesn't fail, when this GC task is processed it will clean up metadata and will enqueue further tasks to cleanup etc in background.